dlrover/python/elastic_agent/torch/ckpt_saver.py

```diff
@@ -409,19 +409,42 @@ def start_async_saving_ckpt(cls):
         from the shared memory into the storage. Firstly, it waits that
         the training process notify the saver class to create a saver.
         """
-        sq = SharedQueue(name="factory", create=True)
 
-        def _save():
-            class_meta: ClassMeta = sq.get()
+        def _saver(class_meta: ClassMeta):
             module = importlib.import_module(class_meta.module_path)
             class_def = getattr(module, class_meta.class_name)
             if cls._saver_instance is None:
                 saver: AsyncCheckpointSaver = class_def(**class_meta.kwargs)
                 cls._saver_instance = saver
             cls._saver_instance._sync_shm_to_storage()
 
+        sq = SharedQueue(name="factory", create=True)
+
+        def _factory():
+            logger.info("Start the checkpoint saver factory.")
+
+            while True:
+                class_meta: ClassMeta = sq.get()
+
+                # use a lock to avoid concurrent creation of the saver
+                with threading.Lock():
+
+                    if cls._saver_instance is not None:
```

Comment:
If the thread to save checkpoint fails, the agent can create a new thread. For example:
```Python
saver_thread  = None
if cls._saver_instance is not None and saver_thread and saver_thread.is_alive():
     continue
saver_thread = threading.Thread(
                        target=_saver,
                        args=(class_meta,),
                        name="checkpoint-saver",
                        daemon=True,
                    )
saver_thread.start()
```